engine/table/src/main/java/io/deephaven/engine/table/impl/remote/ConstructSnapshot.java

```diff
@@ -1613,57 +1749,87 @@ private static <T> Object getSnapshotData(
         }
     }
 
-    private static <T> ArrayList<Chunk<Values>> getSnapshotDataAsChunkList(
-            @NotNull final ColumnSource<T> columnSource,
-            @Nullable final SharedContext sharedContext,
-            @NotNull final RowSet rowSet,
-            final boolean usePrev) {
+    private static void snapshotColumnsSerial(
+            @NotNull final TIntList columnIndices,
+            @NotNull final List<ColumnSource<?>> columnSources,
+            final boolean usePrev,
+            @NotNull final BarrageMessage snapshot) {
+        final RowSet rowSet = snapshot.rowsIncluded;
+        final long numRows = rowSet.size();
+        // The caller should handle empty tables
+        Assert.assertion(numRows > 0, "numRows > 0");
+        final int numCols = columnIndices.size();
         long offset = 0;
-        final long size = rowSet.size();
-        final ArrayList<Chunk<Values>> result = new ArrayList<>();
-
-        if (size == 0) {
-            return result;
-        }
-
-        final int maxChunkSize = (int) Math.min(size, SNAPSHOT_CHUNK_SIZE);
-
-        try (final ColumnSource.FillContext context = columnSource.makeFillContext(maxChunkSize, sharedContext);
+        final int maxChunkSize = (int) Math.min(numRows, SNAPSHOT_CHUNK_SIZE);
+        final ColumnSource.FillContext[] fillContexts = new ColumnSource.FillContext[numCols];
+        try (final SharedContext sharedContext = numCols > 1 ? SharedContext.makeSharedContext() : null;
+                final SafeCloseableArray<ColumnSource.FillContext> ignored = new SafeCloseableArray<>(fillContexts);
                 final RowSequence.Iterator it = rowSet.getRowSequenceIterator()) {
+            for (int colRank = 0; colRank < numCols; ++colRank) {
+                fillContexts[colRank] = columnSources.get(colRank).makeFillContext(maxChunkSize, sharedContext);
+            }
             int chunkSize = maxChunkSize;
             while (it.hasMore()) {
                 final RowSequence reducedRowSet = it.getNextRowSequenceWithLength(chunkSize);
-                final ChunkType chunkType = columnSource.getChunkType();
-
-                // create a new chunk
-                WritableChunk<Values> currentChunk = chunkType.makeWritableChunk(chunkSize);
-
-                if (usePrev) {
-                    columnSource.fillPrevChunk(context, currentChunk, reducedRowSet);
-                } else {
-                    columnSource.fillChunk(context, currentChunk, reducedRowSet);
+                if (reducedRowSet.intSize() != chunkSize) {
+                    failIfConcurrentAttemptInconsistent();
+                    throw new SnapshotUnsuccessfulException(String.format(
+                            "Rowset did not provide requested number of rows in an otherwise consistent state, " +
+                                    "requested = %d, actual = %d",
+                            chunkSize, reducedRowSet.size()));
                 }
```

Comment:
I know we've had some back-and-forth about this code. I don't see how it could actually be hit in practice, since the `snapshot.rowsIncluded` *must* be a copy or intersection owned by the caller. There's no way it should be mutated, even if the table's row set is concurrently being mutated.
```suggestion
```

If you're getting rid of this, maybe we should make this code a little more "classic" by eliminating the offset and chunkSize calculations, and just trust the row set iteration like other code does.